src/lib/core/bidi/directionality.spec.ts

```diff
@@ -0,0 +1,120 @@
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      providers: [Directionality],
+      declarations: [TestAppWithoutDir]
+    });
+
+    TestBed.compileComponents();
+
+    initDir();
+  }));
+
+  it('should read dir from the html element if not specified on the body', () => {
+    document.documentElement.dir = 'rtl';
+
+    let fixture = TestBed.createComponent(TestAppWithoutDir);
+    let testComponent = fixture.debugElement.componentInstance;
+
+    expect(testComponent.dir.value).toBe('rtl');
+  });
+
+  it('should read dir from the body even it is also specified on the html element', () => {
+    document.documentElement.dir = 'ltr';
+    document.body.dir = 'rtl';
+
+    let fixture = TestBed.createComponent(TestAppWithoutDir);
+    let testComponent = fixture.debugElement.componentInstance;
+
+    expect(testComponent.dir.value).toBe('rtl');
+  });
+
+  it('should default to ltr if nothing is specified on either body or the html element', () => {
+    let fixture = TestBed.createComponent(TestAppWithoutDir);
+    let testComponent = fixture.debugElement.componentInstance;
+
+    expect(testComponent.dir.value).toBe('ltr');
+  });
+});
+
+describe('Dir directive', () => {
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [Dir, TestAppWithDir, DirTest]
+    });
+
+    TestBed.compileComponents();
+
+    initDir();
+  }));
+
+  it('should provide itself as Directionality', () => {
+    let fixture = TestBed.createComponent(TestAppWithDir);
+    let testComponent =
+      getDebugNode(fixture.nativeElement.querySelector('dir-test')).componentInstance;
+
+    fixture.detectChanges();
+
+    expect(testComponent.dir.value).toBe('rtl');
+  });
+
+  it('should emit a change event when the value changes', fakeAsync(() => {
+    let fixture = TestBed.createComponent(TestAppWithDir);
+    let testComponent =
+      getDebugNode(fixture.nativeElement.querySelector('dir-test')).componentInstance;
+
+    fixture.detectChanges();
+
+    expect(testComponent.dir.value).toBe('rtl');
+    expect(fixture.componentInstance.changeCount).toBe(0);
+
+    fixture.componentInstance.direction = 'ltr';
+
+    fixture.detectChanges();
+    tick();
+
+    expect(testComponent.dir.value).toBe('ltr');
+    expect(fixture.componentInstance.changeCount).toBe(1);
+  }));
+});
+
+/** Test component without Dir. */
+@Component({
+  selector: 'test-app-without-dir',
+  template: `<div></div>`
+})
+class TestAppWithoutDir {
+  constructor(public dir: Directionality) {}
+}
+
+/** Test component without Dir. */
+@Component({
+  selector: 'test-app-with-dir',
+  template: `
+    <div [dir]="direction" (dirChange)="changeCount= changeCount + 1">
+      <dir-test></dir-test>
+    </div>
+  `
+})
+class TestAppWithDir {
+  direction = 'rtl';
+  changeCount = 0;
+}
+
+/** Test component with Dir directive. */
+@Component({
+  selector: 'dir-test',
+  template: `<div></div>`
+})
+class DirTest {
```

Comment:
Call this `InjectsDirectionality` with selector `injects-directionality`. You can also completely replace `TestAppWithoutDir` with this component.